Andorid 9: panicked with 'java.lang.ClassNotFoundException'

<!-- Thanks for opening an issue! 🤗 -->
<!-- IMPORTANT: This issue tracker is for tracking bugs and features request. For questions or help on how to use Slint, please go to the "Discussions" tab -->
<!-- Please mention your platform and the programming language you are using Slint with -->
<!-- For bugs, please give steps on how to reproduce. What is the expected behavior and what do you see instead. -->
<!-- If possible, please include relevant code snippets, in code blocks (```  ```)-->
Platform: Android
Language: Rust

Reproduce steps:

Project source code: [slint-android-test.zip](https://github.com/slint-ui/slint/files/14735665/slint-android-test.zip)

build project with xbuild (or cargo-apk):
```
x build --arch arm64 --platform android --offline --format apk
```

install and launch APP on Android 9 (with BlueStacks 5 Emulator), will produce error:
```
03-24 21:23:11.649  3864  3884 W System.err: java.lang.ClassNotFoundException: Didn't find class "SlintAndroidJavaHelper" on path: DexPathList[[dex file "InMemoryDexFile[cookie=[0, 127278808428096]]"],nativeLibraryDirectories=[/system/lib64, /system/vendor/lib64]]
03-24 21:23:11.651  3864  3884 W System.err:    at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
03-24 21:23:11.653  3864  3884 W System.err:    Suppressed: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/lang/Object;
03-24 21:23:11.654  3864  3884 W System.err:            at dalvik.system.DexFile.defineClassNative(Native Method)
03-24 21:23:11.654  3864  3884 W System.err:            at dalvik.system.DexFile.defineClass(DexFile.java:283)
03-24 21:23:11.655  3864  3884 W System.err:            at dalvik.system.DexFile.loadClassBinaryName(DexFile.java:276)
03-24 21:23:11.655  3864  3884 W System.err:            at dalvik.system.DexPathList$Element.findClass(DexPathList.java:724)
03-24 21:23:11.655  3864  3884 W System.err:            at dalvik.system.DexPathList.findClass(DexPathList.java:486)
03-24 21:23:11.655  3864  3884 W System.err:            at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:132)
03-24 21:23:11.656  3864  3884 W System.err:    Caused by: java.lang.ClassNotFoundException: Didn't find class "java.lang.Object" on path: DexPathList[[dex file "InMemoryDexFile[cookie=[0, 127278808428096]]"],nativeLibraryDirectories=[/system/lib64, /system/vendor/lib64]]
03-24 21:23:11.657  3864  3884 W System.err:            at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
03-24 21:23:11.660  3864  3884 W System.err:            at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
03-24 21:23:11.662  3864  3884 W System.err:            at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
03-24 21:23:11.663  3864  3884 W System.err:            ... 6 more
03-24 21:23:11.668  3864  3883 I RustStdoutStderr: thread '<unnamed>' panicked at ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/i-slint-backend-android-activity-1.5.1/androidwindowadapter.rs:134:68:
03-24 21:23:11.669  3864  3883 I RustStdoutStderr: JNI error: JavaException
03-24 21:23:11.669  3864  3883 I RustStdoutStderr: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
03-24 21:23:11.675  3864  3884 E RustPanic: JNI error: JavaException
```

same APK works on Andorid 12 (with another emulator and real Android device):
![image](https://github.com/slint-ui/slint/assets/120575400/972f58bd-8bbc-4fe4-89a1-83cb4c210811)

What's the lowest supported Android version at this moment?

Platform: Android 8.0
Slint version: 1.8.0
Probably the same problem. I don't know how to debug the Slint app on Android, but here's the logcat segment:
```
10-03 01:17:25.417 E/AndroidRuntime( 676): at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
10-03 01:17:25.417 E/AndroidRuntime( 676): ... 6 more
10-03 01:17:25.417 E/AndroidRuntime( 676): java.lang.ClassNotFoundException: Didn't find class "SlintAndroidJavaHelper" on path: DexPathList[[dex file "InMemoryDexFile[cookie=[0, 523841617184]]"],nativeLibraryDirectories=[/system/lib64, /system/vendor/lib64]]
10-03 01:17:25.417 E/AndroidRuntime( 676): Process: rust.calculator_rs, PID: 676
10-03 01:17:25.417 E/AndroidRuntime( 676): FATAL EXCEPTION: Thread-3
10-03 01:17:25.417 E/AndroidRuntime( 676): at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:93)
10-03 01:17:25.417 E/AndroidRuntime( 676): Suppressed: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/lang/Object;
10-03 01:17:25.417 E/AndroidRuntime( 676): Caused by: java.lang.ClassNotFoundException: Didn't find class "java.lang.Object" on path: DexPathList[[dex file "InMemoryDexFile[cookie=[0, 523841617184]]"],nativeLibraryDirectories=[/system/lib64, /system/vendor/lib64]]
10-03 01:17:25.417 E/AndroidRuntime( 676): at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
10-03 01:17:25.417 E/AndroidRuntime( 676): at dalvik.system.DexPathList.findClass(DexPathList.java:466)
10-03 01:17:25.417 E/AndroidRuntime( 676): at dalvik.system.DexPathList$Element.findClass(DexPathList.java:677)
10-03 01:17:25.417 E/AndroidRuntime( 676): at dalvik.system.DexFile.loadClassBinaryName(DexFile.java:275)
10-03 01:17:25.417 E/AndroidRuntime( 676): at dalvik.system.DexFile.defineClassNative(Native Method)
10-03 01:17:25.417 E/AndroidRuntime( 676): at dalvik.system.DexFile.defineClass(DexFile.java:282)
10-03 01:17:25.417 E/AndroidRuntime( 676): at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:93)
10-03 01:17:25.417 E/AndroidRuntime( 676): at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:91)
10-03 01:17:25.415 W/System.err( 676): at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
10-03 01:17:25.415 W/System.err( 676): at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
10-03 01:17:25.415 W/System.err( 676): at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:93)
10-03 01:17:25.415 W/System.err( 676): ... 6 more
10-03 01:17:25.415 I/RustStdoutStderr( 676): thread '<unnamed>' panicked at C:\Users\net\.cargo\registry\src\rsproxy.cn-0dccff568467c15b\i-slint-backend-android-activity-1.8.0\androidwindowadapter.rs:171:68:
10-03 01:17:25.415 I/RustStdoutStderr( 676): note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
10-03 01:17:25.415 I/RustStdoutStderr( 676): JNI error: JavaException
10-03 01:17:25.415 E/RustPanic( 676): JNI error: JavaException
10-03 01:17:25.414 W/System.err( 676): at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:93)
10-03 01:17:25.414 W/System.err( 676): Suppressed: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/lang/Object;
10-03 01:17:25.414 W/System.err( 676): Caused by: java.lang.ClassNotFoundException: Didn't find class "java.lang.Object" on path: DexPathList[[dex file "InMemoryDexFile[cookie=[0, 523841617184]]"],nativeLibraryDirectories=[/system/lib64, /system/vendor/lib64]]
10-03 01:17:25.414 W/System.err( 676): at dalvik.system.DexPathList.findClass(DexPathList.java:466)
10-03 01:17:25.414 W/System.err( 676): at dalvik.system.DexPathList$Element.findClass(DexPathList.java:677)
10-03 01:17:25.414 W/System.err( 676): at dalvik.system.DexFile.loadClassBinaryName(DexFile.java:275)
10-03 01:17:25.414 W/System.err( 676): at dalvik.system.DexFile.defineClassNative(Native Method)
10-03 01:17:25.414 W/System.err( 676): at dalvik.system.DexFile.defineClass(DexFile.java:282)
10-03 01:17:25.414 W/System.err( 676): at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:91)
10-03 01:17:25.413 W/System.err( 676): java.lang.ClassNotFoundException: Didn't find class "SlintAndroidJavaHelper" on path: DexPathList[[dex file "InMemoryDexFile[cookie=[0, 523841617184]]"],nativeLibraryDirectories=[/system/lib64, /system/vendor/lib64]]
10-03 01:17:25.412 V/ALMOND ( 676): Almond_Is_DRMDex() : Entering
10-03 01:17:25.352 D/PhoneWindow( 676): windowLightStatusBar : false, disable View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
10-03 01:17:25.352 D/DecorView( 2238): updateNavigationGuardColor navigationGuardColor=0xffe9ebee
10-03 01:17:25.352 D/ViewRootImpl( 2238): onChangedNavigationGuardColor(), color: ffe9ebee
10-03 01:17:25.294 I/zygote64( 676): Increasing code cache capacity to 128KB
10-03 01:17:25.294 I/zygote64( 676): After code cache collection, code=29KB, data=21KB
10-03 01:17:25.294 I/zygote64( 676): Do partial code cache collection, code=29KB, data=21KB
10-03 01:17:25.284 V/APM_AudioPolicyManager( 843): setOutputDevice() setting same device 0x0000 or null device
10-03 01:17:25.284 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 8, device 2
10-03 01:17:25.284 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 7, device 2
10-03 01:17:25.283 V/AudioPolicyManagerEx( 843): getNewOutputDevice() selected device 0
10-03 01:17:25.283 V/APM_AudioPolicyManager( 843): setOutputDevice() prevDevice 0x0002
10-03 01:17:25.283 V/APM_AudioPolicyManager( 843): setOutputDevice() device 0000 delayMs 96
10-03 01:17:25.283 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 6, device 2
10-03 01:17:25.283 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 5, device 2
10-03 01:17:25.283 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 4, device 2
10-03 01:17:25.283 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 3, device 2
10-03 01:17:25.283 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 2, device 2
10-03 01:17:25.283 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 1, device 1
10-03 01:17:25.283 V/APM::AudioPolicyEngine( 843): getDeviceForStrategy() strategy 0, device 2
10-03 01:17:25.283 V/APM::AudioOutputDescriptor( 843): changeRefCount() stream 1, count 0
10-03 01:17:25.283 V/APM_AudioPolicyManager( 843): stopOutput() output 13, stream 1, session 521
10-03 01:17:25.272 V/AudioFlinger_Threads( 843): removeTracks_l removing track on session 521
10-03 01:17:25.269 V/audio_hw_primary( 843): out_get_latency: Latency 8
10-03 01:17:25.256 W/System ( 676): ClassLoader referenced unknown path:
10-03 01:17:25.253 W/ActivityThread( 676): Application rust.calculator_rs can be debugged on port 8100...
```

--- a/internal/backends/android-activity/build.rs
+++ b/internal/backends/android-activity/build.rs
@@ -107,6 +107,8 @@ fn main() {
         .args(&classes)
         .arg("--output")
         .arg(out_dir.as_os_str())
+        .arg("--min-api")
+        .arg("20")
         .output()
         .unwrap_or_else(|err| panic!("Error running {d8_path:?}: {err}"));
 

--- a/internal/backends/android-activity/javahelper.rs
+++ b/internal/backends/android-activity/javahelper.rs
@@ -1,4 +1,4 @@
-﻿// Copyright © SixtyFPS GmbH <info@slint.dev>
+// Copyright © SixtyFPS GmbH <info@slint.dev>
 // SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
 
 use super::*;
@@ -38,18 +38,15 @@ fn load_java_helper(app: &AndroidApp) -> Result<jni::objects::GlobalRef, jni::er
     let parent_dex_loader = env
         .call_method(&native_activity, "getClassLoader", "()Ljava/lang/ClassLoader;", &[])?
         .l()?;
-    
+
     let os_build_class = env.find_class("android/os/Build$VERSION")?;
     let sdk_ver = env.get_static_field(os_build_class, "SDK_INT", "I")?.i()?;
-    
+
     let dex_loader = if sdk_ver >= 26 {
         env.new_object(
             "dalvik/system/InMemoryDexClassLoader",
             "(Ljava/nio/ByteBuffer;Ljava/lang/ClassLoader;)V",
-            &[
-                JValue::Object(&dex_buffer),
-                JValue::Object(&parent_dex_loader)
-            ],
+            &[JValue::Object(&dex_buffer), JValue::Object(&parent_dex_loader)],
         )?
     } else {
         // writes the dex data into the application internal storage
@@ -72,9 +69,8 @@ fn load_java_helper(app: &AndroidApp) -> Result<jni::objects::GlobalRef, jni::er
             "(Ljava/io/File;Ljava/lang/String;)V",
             &[JValue::Object(&dex_dir_path), JValue::Object(&dex_name)],
         )?;
-        let dex_path = env
-            .call_method(dex_path, "getAbsolutePath", "()Ljava/lang/String;", &[])?
-            .l()?;
+        let dex_path =
+            env.call_method(dex_path, "getAbsolutePath", "()Ljava/lang/String;", &[])?.l()?;
 
         // prepares the folder for optimized dex generated while creating `DexClassLoader`
         let out_dex_dir = env.new_string("outdex")?;
@@ -87,30 +83,18 @@ fn load_java_helper(app: &AndroidApp) -> Result<jni::objects::GlobalRef, jni::er
             )?
             .l()?;
         let out_dex_dir_path = env
-            .call_method(
-                &out_dex_dir_path,
-                "getAbsolutePath",
-                "()Ljava/lang/String;",
-                &[],
-            )?
+            .call_method(&out_dex_dir_path, "getAbsolutePath", "()Ljava/lang/String;", &[])?
             .l()?;
 
         // writes the dex data
         let output_stream_class = env.find_class("java/io/FileOutputStream")?;
-        let write_stream = env.new_object(
-            output_stream_class,
-            "(Ljava/lang/String;)V",
-            &[(&dex_path).into()],
-        )?;
+        let write_stream =
+            env.new_object(output_stream_class, "(Ljava/lang/String;)V", &[(&dex_path).into()])?;
         env.call_method(
             &write_stream,
             "write",
             "([BII)V",
-            &[
-                JValue::Object(&dex_byte_array),
-                JValueGen::Int(0),
-                JValueGen::Int(dex_data_len),
-            ],
+            &[JValue::Object(&dex_byte_array), JValueGen::Int(0), JValueGen::Int(dex_data_len)],
         )?;
         env.call_method(&write_stream, "close", "()V", &[])?;
 

--- a/internal/backends/android-activity/javahelper.rs
+++ b/internal/backends/android-activity/javahelper.rs
@@ -1,4 +1,4 @@
-// Copyright © SixtyFPS GmbH <info@slint.dev>
+﻿// Copyright © SixtyFPS GmbH <info@slint.dev>
 // SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
 
 use super::*;
@@ -7,7 +7,7 @@ use i_slint_core::graphics::{euclid, Color};
 use i_slint_core::items::{ColorScheme, InputType};
 use i_slint_core::platform::WindowAdapter;
 use i_slint_core::SharedString;
-use jni::objects::{JClass, JObject, JString, JValue};
+use jni::objects::{JClass, JObject, JString, JValue, JValueGen};
 use jni::sys::{jboolean, jint};
 use jni::JNIEnv;
 use std::time::Duration;
@@ -35,11 +35,97 @@ fn load_java_helper(app: &AndroidApp) -> Result<jni::objects::GlobalRef, jni::er
     let dex_buffer =
         unsafe { env.new_direct_byte_buffer(dex_data.as_ptr() as *mut _, dex_data.len()).unwrap() };
 
-    let dex_loader = env.new_object(
-        "dalvik/system/InMemoryDexClassLoader",
-        "(Ljava/nio/ByteBuffer;Ljava/lang/ClassLoader;)V",
-        &[JValue::Object(&dex_buffer), JValue::Object(&JObject::null())],
-    )?;
+    let parent_dex_loader = env
+        .call_method(&native_activity, "getClassLoader", "()Ljava/lang/ClassLoader;", &[])?
+        .l()?;
+    
+    let os_build_class = env.find_class("android/os/Build$VERSION")?;
+    let sdk_ver = env.get_static_field(os_build_class, "SDK_INT", "I")?.i()?;
+    
+    let dex_loader = if sdk_ver >= 26 {
+        env.new_object(
+            "dalvik/system/InMemoryDexClassLoader",
+            "(Ljava/nio/ByteBuffer;Ljava/lang/ClassLoader;)V",
+            &[
+                JValue::Object(&dex_buffer),
+                JValue::Object(&parent_dex_loader)
+            ],
+        )?
+    } else {
+        // writes the dex data into the application internal storage
+        let dex_data_len = dex_data.len() as i32;
+        let dex_byte_array = env.byte_array_from_slice(dex_data).unwrap();
+
+        let dex_dir = env.new_string("dex")?;
+        let dex_dir_path = env
+            .call_method(
+                &native_activity,
+                "getDir",
+                "(Ljava/lang/String;I)Ljava/io/File;",
+                &[JValue::Object(&dex_dir), JValueGen::Int(0)],
+            )?
+            .l()?;
+        let dex_name = env.new_string(env!("CARGO_CRATE_NAME").to_string() + ".dex")?;
+        let file_class = env.find_class("java/io/File")?;
+        let dex_path = env.new_object(
+            file_class,
+            "(Ljava/io/File;Ljava/lang/String;)V",
+            &[JValue::Object(&dex_dir_path), JValue::Object(&dex_name)],
+        )?;
+        let dex_path = env
+            .call_method(dex_path, "getAbsolutePath", "()Ljava/lang/String;", &[])?
+            .l()?;
+
+        // prepares the folder for optimized dex generated while creating `DexClassLoader`
+        let out_dex_dir = env.new_string("outdex")?;
+        let out_dex_dir_path = env
+            .call_method(
+                &native_activity,
+                "getDir",
+                "(Ljava/lang/String;I)Ljava/io/File;",
+                &[JValue::Object(&out_dex_dir), JValueGen::Int(0)],
+            )?
+            .l()?;
+        let out_dex_dir_path = env
+            .call_method(
+                &out_dex_dir_path,
+                "getAbsolutePath",
+                "()Ljava/lang/String;",
+                &[],
+            )?
+            .l()?;
+
+        // writes the dex data
+        let output_stream_class = env.find_class("java/io/FileOutputStream")?;
+        let write_stream = env.new_object(
+            output_stream_class,
+            "(Ljava/lang/String;)V",
+            &[(&dex_path).into()],
+        )?;
+        env.call_method(
+            &write_stream,
+            "write",
+            "([BII)V",
+            &[
+                JValue::Object(&dex_byte_array),
+                JValueGen::Int(0),
+                JValueGen::Int(dex_data_len),
+            ],
+        )?;
+        env.call_method(&write_stream, "close", "()V", &[])?;
+
+        // loads the dex file
+        env.new_object(
+            "dalvik/system/DexClassLoader",
+            "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/ClassLoader;)V",
+            &[
+                JValue::Object(&dex_path),
+                JValue::Object(&out_dex_dir_path),
+                JValue::Object(&JObject::null()),
+                JValue::Object(&parent_dex_loader),
+            ],
+        )?
+    };
 
     let class_name = env.new_string("SlintAndroidJavaHelper")?;
     let helper_class = env

